### PR TITLE
BREAKING CHANGE: Add generic structure type, use context/getObject

### PR DESCRIPTION
**Pull request by toddbaert**
As described in https://github.com/open-feature/java-sdk/issues/50, I think that expected a `T` back from object flag resolvers is not sustainable. I believe it pushes too much of a burden on providers to instantiate a new T (whatever it may be) and map the attributes received from the backend into it.

Instead, I've added a more generic `Structure` (language used in the OpenFeature spec) with methods similar to the existing `EvaluationContext`. I've also made the `EvaluationContext` inherit from this to reduce duplication (the `EvaluationContext` is also referred to as a "structure" in the spec. 

After this change, providers will create a new `Structure` and add the corresponding fields from the json/xml/protobuf they are concerned with.

In summary:

- Adds `Structure` type which represents JSON data
- `EvaluationContext` ~inherits~ encapsultes this using lombok's `@Delegate`.
- Consolidated `addIntegerAttribute`, `addStringAttribute`, etc, in `EvaluationContext` to a single overloaded `add` method.
- Removed ability to get multiple attributes of a single type from `EvaluationContext` (I'm not convinced these are that valuable, but I'll take feedback).
- introduces a `Value` wrapper class that houses one of: Structure, List<Value>, Boolean, etc (representations of all the possible JSON node values).

### CODE REVIEW

**Comment by beeme1mr** [comment]
FYI @rgrassian-split

**Comment by rgrassian-split** [comment]
> FYI @rgrassian-split. I would like to get your thoughts on this. How would this change impact the provider you're working on?

@beeme1mr Hey! This is a fantastic change, I was actually running into some issues with the `T` implementation so I'm happy to see this. It shouldn't be too crazy to change our provider to support this.

**Comment by toddbaert** [comment]
@rgrassian-split it would be good if you could put some feedback in https://github.com/open-feature/java-sdk/issues/50

**Comment by rgrassian-split** [comment]
@toddbaert Just did, that is actually the bigger deal for us that I very much support switching it from the current approach

File: pom.xml
Comment (justinabrahms) [inline pom.xml:1]: If you would like to change whitespace in this file (or others), please check in a tool which will break the build if it's not right. I'm not tied to my particular formatting choices, but don't want us to fight back and forth about what the right answer should be.

File: pom.xml
Comment (justinabrahms) [inline pom.xml:248]: Reminder that we'll want to uncomment these here and elsewhere.

File: src/main/java/dev/openfeature/javasdk/EvaluationContext.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/EvaluationContext.java:4]: Similar to above, either make your editor not do these formatting changes or check in something that consistently applies it.

File: src/main/java/dev/openfeature/javasdk/EvaluationContext.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/EvaluationContext.java:14]: I'd actually prefer we not split this across 3 lines. `@Setter` isn't enough information to have it's own line, imo.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/Structure.java:52]: Unrelated to your change, but I'm starting to think we should drop this. It feels like it should migrate to contrib, at best.

File: src/test/java/dev/openfeature/javasdk/EvalContextTest.java
Comment (justinabrahms) [inline src/test/java/dev/openfeature/javasdk/EvalContextTest.java:124]: You dropped this one from your test.

**Review by justinabrahms — CHANGES_REQUESTED** [review]
Lots of formatting churn makes the testing stuff (especially spec definitions) a bit hard to read.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/Structure.java:60]: We're also missing something like an array type.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/Structure.java:16]: If our backend returns this:

```json
{
  "foo": null
}
```

What type will that serialize into on a structure?

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:52]: I think I agree... but there's something like this in the spec. Do you think we need to adjust the spec?

File: pom.xml
Comment (toddbaert) [reply pom.xml:248]: ugh. I didn't mean to commit this.

I have an issue with this pom and Java 18. I was able to resolve it in the contribs.

Comment (toddbaert) [reply pom.xml]: reverted.

File: pom.xml
Comment (toddbaert) [reply pom.xml]: I will see what I can do about this.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:60]: I think you are right.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/Structure.java:16]: Provider won't know what type to use. I think the way we work around this is returning `null` for any unknown values. This means if you come across a `null`, you can just not call `add` at all.

File: src/test/java/dev/openfeature/javasdk/EvalContextTest.java
Comment (toddbaert) [reply src/test/java/dev/openfeature/javasdk/EvalContextTest.java:124]: Added back, thanks!

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java]: Added

**Comment by toddbaert** [comment]
@justinabrahms I think I've cleaned up a lot of the whitespace changes. A few are possibly improvements (missing newlines in some places, etc). I've left those.

I also added array support in `Structure`.

I'm going to test this locally in my provider and see if it makes things nicer. Let me know what you think.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:17]: > If we don't call add, then foo won't be in the keySet. So... let's just pick something like boolean and shove all nulls in there as a typecasted thing. Users will get nulls out when they ask for them.

OK that makes sense. That's really something for the provider to do though, right? I guess we could add tests for this, but there isn't much for us to do in the SDK to support this, right?

Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/Structure.java:17]: That's right.

Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:17]: I'll add a specific test for this.

**Comment by justinabrahms** [comment]
Playing around with the API. It seems like because we broke out `Structure`, the `return this` is returning `Structure` objects, not `EvaluationContext` objects. I'm not aware of a way to have it return `EvaluationContext` objects except to have EC proxy those methods to an internal representation of a `Structure`. Thoughts?

image

**Comment by toddbaert** [comment]
> Playing around with the API. It seems like because we broke out `Structure`, the `return this` is returning `Structure` objects, not `EvaluationContext` objects. I'm not aware of a way to have it return `EvaluationContext` objects except to have EC proxy those methods to an internal representation of a `Structure`. Thoughts?

I noticed that too... it's sort of a shame, but I think we'd have to proxy them as well... maybe there's a clever solution but I can't think of one immediately. I'll look around.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:17]: I've added such a test.

File: .github/workflows/pullrequest.yml
Comment (toddbaert) [inline .github/workflows/pullrequest.yml:43]: Removed signing in PRs.

File: pom.xml
Comment (toddbaert) [inline pom.xml:296]: Only sign in deploy. Nice because now you can run `mvn verify` locally and in PRs without worrying about GPG setup.

File: src/main/java/dev/openfeature/javasdk/EvaluationContext.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/EvaluationContext.java:13]: I'm thinking about whether we should be extending structure here, rather than having an instance of Structure as an internal property. I think I have a preference for the latter?

One thing that makes me nervous about this EC/Structure merging is that Array type you added to the flag value types. We don't actually support returning arrays, and so parts of those are leaking between one another.

Thoughts?

Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/EvaluationContext.java:13]: Splitting out the enum will address my concern for FVT. `@Delegate` looks like exactly what I was hoping for. Lombok is the best. Thank you!

Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/EvaluationContext.java:13]: Lombok is incredible.

Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/EvaluationContext.java]: I believe [this commit](https://github.com/open-feature/java-sdk/pull/51/commits/a8623253547411e94fe457adbb1fdd1a973779d7) resolves this concerns.

- encapsulates instead of inherits `Structure`, using lombok's `@Delegate`.
- adds new internal enum for `Structure` field types.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (kinyoklion) [inline src/main/java/dev/openfeature/javasdk/Structure.java:119]: Would it be possible to get something like this that includes the type annotations? So that if one is converting this structure into something using another structure, or a builder, that they can iterate it and dispatch the methods to the correct types?

Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/Structure.java:119]: @kinyoklion I think you should be able to do that with `instanceof`?

**Review by justinabrahms — APPROVED** [review]

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:119]: > @kinyoklion I think you should be able to do that with instanceof?

Indeed, this is what I'm doing in my provider. Is there some reason it's better to leverage the underlying annotation in the `Pair`?

Comment (kinyoklion) [reply src/main/java/dev/openfeature/javasdk/Structure.java:119]: Yeah, it can be done that way. Also the performance of `instanceof` seems to be good.

The information was already there, so having to immediately re-construct it, seemed a bit counter intuitive.

Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/Structure.java:119]: Totally. I just don't want it to be part of a public API

**Comment by toddbaert** [comment]
@justinabrahms I'd like to play with this a bit more locally before we merge... I'll let you know if I think any changes are needed. I think this should be sufficient though.

**Comment by toddbaert** [comment]
@justinabrahms  I made some more changes. Our abstraction was leaky because any node in a JSON structure can be a JSON structure, an array or a primitive. That wasn't reflected the implementation.

What I've done in [this commit](https://github.com/open-feature/java-sdk/pull/51/commits/ee58654f6bc02fa5e31b0c734fb85a6068e3cda3), is introduce a `Value` wrapper class that houses one of: `Structure`, `List<Value>`, `Boolean`, etc (representations of all the possible JSON node values). The actual value can be retrieved like `value.asBoolean()` or `value.asStructure()`. I believe any JSON structure can be represented this way.

This is similar to how java-grpc represents their structure (which is very similar to JSON) and something like a simplified version of LDValue.

I've maintained most of the previous work. `EvaluationContext` still encapsulates a `Structure`, etc.

File: src/main/java/dev/openfeature/javasdk/EvaluationContext.java
Comment (toddbaert) [inline src/main/java/dev/openfeature/javasdk/EvaluationContext.java:89]: Lists must be of type `Value`.

File: src/main/java/dev/openfeature/javasdk/FlagEvaluationOptions.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/FlagEvaluationOptions.java:10]: Not clear why this is being qualified rather than importing it as above? Is it because it's only used once?

File: src/main/java/dev/openfeature/javasdk/Value.java
Comment (kinyoklion) [inline src/main/java/dev/openfeature/javasdk/Value.java:105]: Because it shares a storage format with string, would it be worthwhile to have a comment that dates are also strings? If you do care about dates, and you are using this in some kind of dispatch, then you would want to check isZonedDateTime before isString, or check both, or something to disambiguate.

File: src/main/java/dev/openfeature/javasdk/Value.java
Comment (justinabrahms) [inline src/main/java/dev/openfeature/javasdk/Value.java:13]:
```suggestion
public class StructureValue {
```
?

**Review by justinabrahms — COMMENTED** [review]
I think I need to play around with constructing a few of these objects to figure out what the API feels like.

File: src/main/java/dev/openfeature/javasdk/FlagEvaluationOptions.java
Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/FlagEvaluationOptions.java:10]: NM. It's now ambiguous. I get it.

File: src/main/java/dev/openfeature/javasdk/internal/StructureFieldType.java
Comment (kinyoklion) [inline src/main/java/dev/openfeature/javasdk/internal/StructureFieldType.java:3]: This isn't used now that there pairs were removed?

**Review by kinyoklion — APPROVED** [review]

File: src/main/java/dev/openfeature/javasdk/Value.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Value.java:13]: Not sure I agree with this one. I would be OK with `Node`, but I think `Value` is better.

Adding "Structure" seems wrong, since the value can be a primitive.

File: src/main/java/dev/openfeature/javasdk/internal/StructureFieldType.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/internal/StructureFieldType.java:3]: Yep, great catch.

Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/internal/StructureFieldType.java:3]: Removed.

**Comment by toddbaert** [comment]
> I think I need to play around with constructing a few of these objects to figure out what the API feels like.

OK @justinabrahms, I'll leave it up to you to merge it then (or message with more feedback). @kinyoklion has also approved. I think we'll need to do something similar in dotnet cc @benjiro

File: src/main/java/dev/openfeature/javasdk/Value.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Value.java:105]: I'm wondering if it isn't better to simply store a datetime... Storing them as strings only adds this complication, as well as performance issues around parsing/error handling. It was stored as a string in the previous implementation and I didn't give it much thought when I translated it.

Comment (kinyoklion) [reply src/main/java/dev/openfeature/javasdk/Value.java:105]: I think storing it as a datetime, or removing it, are both good options.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (s-sen) [inline src/main/java/dev/openfeature/javasdk/Structure.java:85]: Are we OK with a shallow copy of the values? The application developer will use this to obtain all the values hence a  shallow copy is OK in that scenario. But will this structure also be used during building and mutating the EvaluationContext?

Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Structure.java:85]: I think this is a good question...

I suspect that for mutating, users will be using the `merge` and `add` functions in the EvaluationContext. I don't have any strong opinions about if this should be a shallow copy, deep copy, or just expose the inner map (though I think in most cases, that's a bad choice). I guess whatever we choose it should be well-documented.

Does anyone have a strong argument for this?

@justinabrahms @kinyoklion

File: src/main/java/dev/openfeature/javasdk/Value.java
Comment (toddbaert) [reply src/main/java/dev/openfeature/javasdk/Value.java:103]: Now storing as ZonedDateTIme.

File: src/main/java/dev/openfeature/javasdk/Structure.java
Comment (justinabrahms) [reply src/main/java/dev/openfeature/javasdk/Structure.java:92]: https://github.com/open-feature/java-sdk/issues/53 Moved this discussion here.

**Pull request merged**
